examples/virtual-dom-wip/platform/Html/HostJavaScript.roc

```diff
@@ -0,0 +1,279 @@
+interface Html.HostJavaScript
+    exposes [hostJavaScript]
+    imports []
+
+hostJavaScript : Str
+hostJavaScript =
+    """
+    /** @typedef {(e: Event) => void} JsEventDispatcher */
+
+    /** @typedef {[string, JsEventDispatcher]} Listener */
+
+    /**
+     * @typedef {Object} RocWasmExports
+     * @property {(size: number, alignment: number) => number} roc_alloc
+     * @property {(jsonListAddr: number, jsonListLength: number, handlerId: number) => void} roc_dispatch_event
+     * @property {() => number} main
+     */
+
+    /**
+     * @typedef {{ObjectField: [string, CyclicStructureAccessor]} | {ArrayIndex: [number, CyclicStructureAccessor]} | {SerializableValue: []}} CyclicStructureAccessor
+     */
+
+    /** @type {Array<Node | null>} */
+    const nodes = [];
+
+    /** @type {Array<Listener | null>} */
+    const listeners = [];
+
+    const utf8Decoder = new TextDecoder();
+    const utf8Encoder = new TextEncoder();
+
+    /**
+     * @param {string} wasmFilename
+     */
+    export const init = async (wasmFilename) => {
+      const effects = {
+        /**
+         * @param {number} tagAddr
+         */
+        createElement: (tagAddr) => {
+          const tagName = decodeRocStr(tagAddr);
+          const node = document.createElement(tagName);
+          return insertNode(node);
+        },
+
+        /**
+         * @param {number} contentAddr
+         */
+        createTextNode: (contentAddr) => {
+          const content = decodeRocStr(contentAddr);
+          const node = document.createTextNode(content);
+          return insertNode(node);
+        },
+
+        /**
+         * @param {number} parentId
+         * @param {number} childId
+         */
+        appendChild: (parentId, childId) => {
+          const parent = nodes[parentId];
+          const child = nodes[childId];
+          parent.appendChild(child);
+        },
+
+        /**
+         * @param {number} id
+         */
+        removeNode: (id) => {
+          var _a;
```

Comment:
Is this var necessary?

Reply:
Doesn't look like it, no! Thanks for spotting.
I wrote the earlier versions in TypeScript, then generated the JS. The TypeScript compiler does stuff like this sometimes. I'm sure there was a good reason but the source might have changed. Not great that I'm keeping both a host.js and this file. But when it stabilises a bit more I will get rid of one.